core/src/main/scala/org/apache/spark/sql/execution/columnar/impl/ColumnFormatRelation.scala

```diff
@@ -478,20 +633,55 @@ final class DefaultSource extends ColumnarRelationProvider {
       schema, externalStore)
 
     var success = false
-    val relation = new ColumnFormatRelation(SnappyStoreHiveCatalog.
+
+    // create an index relation if it is an index table
+    val relation = options.get(StoreUtils.GEM_INDEXED_TABLE) match {
+      case Some(baseTable) => new IndexColumnFormatRelation(SnappyStoreHiveCatalog.
+        processTableIdentifier(table, sqlContext.conf),
+        getClass.getCanonicalName,
+        mode,
+        schema,
+        schemaExtension,
+        ddlExtensionForShadowTable,
+        options,
+        externalStore,
+        blockMap,
+        partitioningColumn,
+        sqlContext,
+        baseTable)
+      case None => new ColumnFormatRelation(SnappyStoreHiveCatalog.
         processTableIdentifier(table, sqlContext.conf),
-      getClass.getCanonicalName,
-      mode,
-      schema,
-      schemaExtension,
-      ddlExtensionForShadowTable,
-      options,
-      externalStore,
-      blockMap,
-      partitioningColumn,
-      sqlContext)
+        getClass.getCanonicalName,
+        mode,
+        schema,
+        schemaExtension,
+        ddlExtensionForShadowTable,
+        options,
+        externalStore,
+        blockMap,
+        partitioningColumn,
+        sqlContext)
+    }
+
     try {
       relation.createTable(mode)
+      if (relation.isInstanceOf[ColumnFormatRelation]) {
+        var indexes: Array[String] = Array()
+        try {
+          val options = new CaseInsensitiveMutableHashMap(relation.origOptions)
+          indexes = options(ExternalStoreUtils.INDEX_NAME).split(",")
+        } catch {
+          case e: NoSuchElementException =>
+        }
+
+        indexes.foreach(index => {
+          val sncCatalog = sqlContext.asInstanceOf[SnappyContext].catalog
+          val dr = sncCatalog.lookupRelation(sncCatalog.newQualifiedTableName(index)) match {
+            case LogicalRelation(r: DependentRelation, _) => r
+          }
+          relation.asInstanceOf[ColumnFormatRelation].addDependent(dr, sncCatalog)
+        })
+      }
```

Comment:
Previous comment that got folded as outdated.

We should look at adding generic dependency management rather than specific for index (and previous stream tables). For this case it can look like:

```
// relation recovered that implements ParentRelation (in cachedDataSourceTables CacheLoader.load)
// when ParentRelation is stored in hive catalog then it should have a generic property like "DEPENDENTS" that contains all the dependent relation names with type names (e.g. INDEX can be a type)
// the dependent types will be an enum in Scala code but in hive catalog its name will be stored like done for ExternalTableType)
// for each of the retrieved dependents, add to DependencyCatalog (the dependency type can also be added to DependencyCatalog)
```

If you don't want to burden this PR with above, then please track as a JIRA.